dbms/src/Core/SettingsCommon.h

```diff
@@ -415,6 +425,18 @@ class SettingsCollection
         reference(const const_reference & src) : const_reference(src) {}
         void setValue(const Field & value) { this->member->set_field(*const_cast<Derived *>(this->collection), value); }
         void setValue(const String & value) { this->member->set_string(*const_cast<Derived *>(this->collection), value); }
+        void updateValue(const Field & value)
+        {
+            if (!this->member->updateable)
+                throw Exception("Setting '" + this->member->name.toString() + "' is restricted for updates.", ErrorCodes::IMMUTABLE_SETTING);
+            setValue(value);
+        }
```

Comment:
Looks too complicated. Why do we need to distinguish mutable settings and immutable settings? Why can't they all be mutable?

Reply:
Because we shouldn't be able to change some of them, for example, index_granularity for MergeTree.

Reply:
I believe it's better to check these conditions externally because I'd prefer to keep the class template `SettingsCollection` as simply as possible. Besides we already have the class `SettingsConstraints` which makes very similar checks, and it doesn't seem good to have two completely different mechanisms for the same thing. Maybe you can implement something like MergeTreeSettingsConstraings instead?

Reply:
From my point of view, updateability is a basic property of each setting. It should be defined as part of each setting. Maybe one more macro is not a good solution, but I don't see big complexity in additional `const bool` flag.